File: apps/android/app/src/main/java/com/example/aischoolplatform/dev/parent/service/ParentAppService.kt
package com.example.aischoolplatform.dev.parent.service

import com.example.aischoolplatform.dev.parent.data.AnnouncementRepository
import com.example.aischoolplatform.dev.parent.data.AppPreferenceRepository
import com.example.aischoolplatform.dev.parent.data.FormRepository
import com.example.aischoolplatform.dev.parent.data.ParentRepository
import com.example.aischoolplatform.dev.parent.model.ChildSummary
import com.example.aischoolplatform.dev.parent.model.LanguagePreference
import com.example.aischoolplatform.dev.parent.model.ParentAnnouncement
import com.example.aischoolplatform.dev.parent.model.ParentFormAnswer
import com.example.aischoolplatform.dev.parent.model.ParentFormQuestionType
import com.example.aischoolplatform.dev.parent.model.ParentFormStatus
import com.example.aischoolplatform.dev.parent.model.ParentFormTask
import com.example.aischoolplatform.dev.parent.model.ParentHomeState
import com.example.aischoolplatform.dev.parent.model.ParentRole
import com.example.aischoolplatform.dev.parent.model.ParentSession
import java.time.Instant

sealed class ParentAppResult<out T> {
    data class Success<T>(val value: T) : ParentAppResult<T>()
    data class Failure(val message: String) : ParentAppResult<Nothing>()
}

class ParentAppService(
    private val parentRepository: ParentRepository,
    private val announcementRepository: AnnouncementRepository,
    private val formRepository: FormRepository,
    private val appPreferenceRepository: AppPreferenceRepository,
    private val now: () -> Instant = { Instant.now() }
) {
    fun currentSession(): ParentSession? = parentRepository.currentSession()

    fun homeState(session: ParentSession, selectedChildId: String? = null): ParentAppResult<ParentHomeState> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        val parent = parentRepository.parentProfile(session) ?: return ParentAppResult.Failure("Parent profile was not found.")
        val school = parentRepository.school(session) ?: return ParentAppResult.Failure("School was not found.")
        val children = parentRepository.linkedChildren(session)
        val selectedChild = selectedChildId?.let { parentRepository.child(session, it) } ?: children.firstOrNull()
        if (selectedChildId != null && selectedChild == null) return ParentAppResult.Failure("That child is not linked to this parent account.")
        return ParentAppResult.Success(
            ParentHomeState(
                parent = parent,
                school = school,
                selectedChild = selectedChild,
                children = children,
                announcements = announcementRepository.publishedAnnouncements(session),
                forms = formRepository.formTasks(session)
            )
        )
    }

    fun linkedChildren(session: ParentSession): ParentAppResult<List<ChildSummary>> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        return ParentAppResult.Success(parentRepository.linkedChildren(session))
    }

    fun child(session: ParentSession, studentId: String): ParentAppResult<ChildSummary> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        val child = parentRepository.child(session, studentId)
            ?: return ParentAppResult.Failure("That child is not linked to this parent account.")
        return ParentAppResult.Success(child)
    }

    fun inbox(session: ParentSession): ParentAppResult<List<ParentAnnouncement>> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        return ParentAppResult.Success(announcementRepository.publishedAnnouncements(session))
    }

    fun announcement(session: ParentSession, announcementId: String): ParentAppResult<ParentAnnouncement> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        val announcement = announcementRepository.announcement(session, announcementId)
            ?: return ParentAppResult.Failure("Announcement was not found for this parent account.")
        return ParentAppResult.Success(announcement)
    }

    fun markAnnouncementRead(session: ParentSession, announcementId: String): ParentAppResult<ParentAnnouncement> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        val updated = announcementRepository.markRead(session, announcementId, now().toString())
            ?: return ParentAppResult.Failure("Announcement was not found for this parent account.")
        return ParentAppResult.Success(updated)
    }

    fun forms(session: ParentSession): ParentAppResult<List<ParentFormTask>> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        return ParentAppResult.Success(formRepository.formTasks(session))
    }

    fun form(session: ParentSession, recipientId: String): ParentAppResult<ParentFormTask> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        val form = formRepository.formTask(session, recipientId)
            ?: return ParentAppResult.Failure("Form was not found for this parent account.")
        return ParentAppResult.Success(form)
    }

    fun submitForm(
        session: ParentSession,
        recipientId: String,
        answers: List<ParentFormAnswer>
    ): ParentAppResult<ParentFormTask> {
        if (!isParentSession(session)) return ParentAppResult.Failure("Parent access is not available for this user.")
        val form = formRepository.formTask(session, recipientId)
            ?: return ParentAppResult.Failure("Form was not found for this parent account.")
        if (form.status != ParentFormStatus.Outstanding) return ParentAppResult.Failure("This form cannot be submitted.")
        if (deadlineHasPassed(form.deadlineAt)) return ParentAppResult.Failure("The deadline for this form has passed.")
        val validation = validateAnswers(form, answers)
        if (validation != null) return ParentAppResult.Failure(validation)
        val updated = formRepository.submitForm(session, recipientId, answers, now().toString())
            ?: return ParentAppResult.Failure("This form could not be submitted.")
        return ParentAppResult.Success(updated)
    }

    fun languagePreference(): LanguagePreference = appPreferenceRepository.getLanguage()

    fun setLanguagePreference(preference: LanguagePreference): LanguagePreference {
        appPreferenceRepository.setLanguage(preference)
        return appPreferenceRepository.getLanguage()
    }

    private fun deadlineHasPassed(deadlineAt: String?): Boolean =
        deadlineAt?.let { runCatching { !Instant.parse(it).isAfter(now()) }.getOrDefault(true) } ?: false

    private fun validateAnswers(form: ParentFormTask, answers: List<ParentFormAnswer>): String? {
        val answersByQuestion = answers.associateBy { it.questionId }
        form.questions.filter { it.required }.forEach { question ->
            val answer = answersByQuestion[question.id]?.value?.trim()
            if (answer.isNullOrEmpty()) return "Required form questions must be answered."
            if ((question.type == ParentFormQuestionType.Acknowledgement || question.type == ParentFormQuestionType.Consent) && answer != "true") {
                return "Consent and acknowledgement questions must be explicitly accepted."
            }
        }
        return null
    }

    private fun isParentSession(session: ParentSession): Boolean =
        session.role == ParentRole.ParentGuardian
}

File: apps/ios/App/Parent/Services/ParentAppService.swift
import Foundation

enum ParentAppError: Error, LocalizedError, Equatable {
    case notParent
    case parentNotFound
    case schoolNotFound
    case childNotLinked
    case announcementNotFound
    case formNotFound
    case formNotSubmittable
    case formDeadlinePassed
    case requiredFormAnswersMissing
    case consentNotAccepted

    var errorDescription: String? {
        switch self {
        case .notParent:
            return "Parent access is not available for this user."
        case .parentNotFound:
            return "Parent profile was not found."
        case .schoolNotFound:
            return "School was not found."
        case .childNotLinked:
            return "That child is not linked to this parent account."
        case .announcementNotFound:
            return "Announcement was not found for this parent account."
        case .formNotFound:
            return "Form was not found for this parent account."
        case .formNotSubmittable:
            return "This form cannot be submitted."
        case .formDeadlinePassed:
            return "The deadline for this form has passed."
        case .requiredFormAnswersMissing:
            return "Required form questions must be answered."
        case .consentNotAccepted:
            return "Consent and acknowledgement questions must be explicitly accepted."
        }
    }
}

final class ParentAppService {
    private let parentRepository: ParentRepository
    private let announcementRepository: ParentAnnouncementRepository
    private let formRepository: ParentFormRepository
    private let now: () -> Date

    init(
        parentRepository: ParentRepository,
        announcementRepository: ParentAnnouncementRepository,
        formRepository: ParentFormRepository,
        now: @escaping () -> Date = Date.init
    ) {
        self.parentRepository = parentRepository
        self.announcementRepository = announcementRepository
        self.formRepository = formRepository
        self.now = now
    }

    func currentSession() -> ParentSession? {
        parentRepository.currentSession()
    }

    func homeState(session: ParentSession, selectedChildId: String? = nil) throws -> ParentHomeState {
        try requireParent(session)
        guard let parent = parentRepository.parentProfile(for: session) else { throw ParentAppError.parentNotFound }
        guard let school = parentRepository.school(for: session) else { throw ParentAppError.schoolNotFound }
        let children = parentRepository.linkedChildren(for: session)
        let selectedChild = try selectedChildId.map { id in
            guard let child = parentRepository.child(for: session, studentId: id) else { throw ParentAppError.childNotLinked }
            return child
        } ?? children.first
        return ParentHomeState(
            parent: parent,
            school: school,
            selectedChild: selectedChild,
            children: children,
            announcements: announcementRepository.publishedAnnouncements(for: session),
            forms: formRepository.formTasks(for: session)
        )
    }

    func linkedChildren(session: ParentSession) throws -> [ChildSummary] {
        try requireParent(session)
        return parentRepository.linkedChildren(for: session)
    }

    func child(session: ParentSession, studentId: String) throws -> ChildSummary {
        try requireParent(session)
        guard let child = parentRepository.child(for: session, studentId: studentId) else { throw ParentAppError.childNotLinked }
        return child
    }

    func inbox(session: ParentSession) throws -> [ParentAnnouncement] {
        try requireParent(session)
        return announcementRepository.publishedAnnouncements(for: session)
    }

    func announcement(session: ParentSession, announcementId: String) throws -> ParentAnnouncement {
        try requireParent(session)
        guard let announcement = announcementRepository.announcement(for: session, announcementId: announcementId) else { throw ParentAppError.announcementNotFound }
        return announcement
    }

    func markAnnouncementRead(session: ParentSession, announcementId: String) throws -> ParentAnnouncement {
        try requireParent(session)
        let timestamp = ISO8601DateFormatter().string(from: now())
        guard let announcement = announcementRepository.markRead(for: session, announcementId: announcementId, readAt: timestamp) else { throw ParentAppError.announcementNotFound }
        return announcement
    }

    func forms(session: ParentSession) throws -> [ParentFormTask] {
        try requireParent(session)
        return formRepository.formTasks(for: session)
    }

    func form(session: ParentSession, recipientId: String) throws -> ParentFormTask {
        try requireParent(session)
        guard let task = formRepository.formTask(for: session, recipientId: recipientId) else { throw ParentAppError.formNotFound }
        return task
    }

    func submitForm(session: ParentSession, recipientId: String, answers: [ParentFormAnswer]) throws -> ParentFormTask {
        try requireParent(session)
        let task = try form(session: session, recipientId: recipientId)
        guard task.status == .outstanding else { throw ParentAppError.formNotSubmittable }
        guard !deadlineHasPassed(task.deadlineAt) else { throw ParentAppError.formDeadlinePassed }
        try validateAnswers(task: task, answers: answers)
        let timestamp = ISO8601DateFormatter().string(from: now())
        guard let updated = formRepository.submitForm(for: session, recipientId: recipientId, answers: answers, submittedAt: timestamp) else { throw ParentAppError.formNotSubmittable }
        return updated
    }

    private func deadlineHasPassed(_ deadlineAt: String?) -> Bool {
        guard let deadlineAt, let deadline = ISO8601DateFormatter().date(from: deadlineAt) else { return deadlineAt != nil }
        return deadline <= now()
    }

    private func validateAnswers(task: ParentFormTask, answers: [ParentFormAnswer]) throws {
        let answersByQuestion = Dictionary(uniqueKeysWithValues: answers.map { ($0.questionId, $0.value.trimmingCharacters(in: .whitespacesAndNewlines)) })
        for question in task.questions where question.required {
            guard let answer = answersByQuestion[question.id], !answer.isEmpty else { throw ParentAppError.requiredFormAnswersMissing }
            if [.acknowledgement, .consent].contains(question.type), answer != "true" {
                throw ParentAppError.consentNotAccepted
            }
        }
    }

    private func requireParent(_ session: ParentSession) throws {
        guard session.role == .parentGuardian else { throw ParentAppError.notParent }
    }
}

enum DevelopmentParentComposition {
    static func createService() -> ParentAppService {
        let repository = DevelopmentParentRepository()
        return ParentAppService(parentRepository: repository, announcementRepository: repository, formRepository: repository)
    }
}